Mathlib/Data/UInt.lean

```diff
@@ -30,73 +24,113 @@ version also interferes more with software-verification use-cases, which is reas
---         natCast n := ofBitVec n
-
---       lemma natCast_def (n : ℕ) : (n : $typeName) = ofBitVec n := rfl
-
---       lemma intCast_def (z : ℤ) : (z : $typeName) = ofBitVec z := rfl
-
---       local instance instCommRing : CommRing $typeName :=
---         Function.Injective.commRing toBitVec toBitVec_injective
---           rfl rfl (fun _ _ => rfl) (fun _ _ => rfl) (fun _ => rfl) (fun _ _ => rfl)
---           (fun _ _ => rfl) (fun _ _ => rfl) (fun _ _ => rfl) (fun _ => rfl) (fun _ => rfl)
-
---       namespace CommRing
---       attribute [scoped instance] instCommRing instNatCast instIntCast
---       end CommRing
-
---     end $typeName
---   ))
---   -- interpolating docstrings above is more trouble than it's worth
---   let docString :=
---     s!"To use this instance, use `open scoped {typeName'}.CommRing`.\n\n" ++
---     "See the module docstring for an explanation"
---   Lean.addDocStringCore (typeName'.mkStr "instCommRing") docString
---   Lean.addDocStringCore (typeName'.mkStr "instNatCast") docString
---   Lean.addDocStringCore (typeName'.mkStr "instIntCast") docString
+-- these theorems are fragile, so do them first
+set_option hygiene false in
+run_cmd
+  for typeName' in [`UInt8, `UInt16, `UInt32, `UInt64, `USize] do
+  let typeName := Lean.mkIdent typeName'
+  Lean.Elab.Command.elabCommand (← `(
+    namespace $typeName
+
+      open $typeName (toBitVec_mul)
+      @[simp, int_toBitVec]
+      protected theorem toBitVec_pow (a : $typeName) (n : ℕ) :
+          (a ^ n).toBitVec = a.toBitVec ^ n := by
+        induction n with
+        | zero => simp
+        | succ n ih =>
+          change (a ^ n * a).toBitVec = _
+          rw [toBitVec_mul, ih]
+          rfl
+
+      attribute [local instance] natCast intCast
+
+      @[simp, int_toBitVec]
+      protected theorem toBitVec_natCast (n : ℕ) :
+          (n : $typeName).toBitVec = n := rfl
+
+      open $typeName (toBitVec_neg) in
+      @[simp, int_toBitVec]
+      protected theorem toBitVec_intCast (z : ℤ) :
+          (z : $typeName).toBitVec = z := by
+        obtain ⟨z, rfl | rfl⟩ := z.eq_nat_or_neg
+        · erw [intCast_ofNat]; rfl
+        · rw [intCast_neg, toBitVec_neg]
+          erw [intCast_ofNat]
+          simp
+
+    end $typeName
+  ))
+
+set_option hygiene false in
+run_cmd
+  for typeName' in [`UInt8, `UInt16, `UInt32, `UInt64, `USize] do
+  let typeName := Lean.mkIdent typeName'
+  Lean.Elab.Command.elabCommand (← `(
+    namespace $typeName
+
+      instance : SMul ℕ $typeName where
+        smul n a := ofBitVec ⟨n • a.toFin⟩
+
+      @[simp, int_toBitVec]
+      protected theorem toBitVec_nsmul (n : ℕ) (a : $typeName) :
+          (n • a).toBitVec = n • a.toBitVec := rfl
+
+      instance : SMul ℤ $typeName where
+        smul z a := ofBitVec ⟨z • a.toFin⟩
+
+      @[simp, int_toBitVec]
+      protected theorem toBitVec_zsmul (z : ℤ) (a : $typeName) :
+          (z • a).toBitVec = z • a.toBitVec := rfl
```

Comment:
I've just noticed these instances create diamonds. I'll remove them, which will mean this file creates no new data overall.